src/can.rs

```diff
@@ -0,0 +1,498 @@
+        };
+
+        return (transmitter, fifo0, fifo1);
+    }
+
+    pub fn release(self) -> stm32::CAN {
+        self.can
+    }
+}
+
+impl embedded_hal_can::Interface for CanTransmitter {
+    type Id = CanId;
+    type Frame = CanFrame;
+    type Error = ();
+    type Filter = CanFilter;
+}
+
+impl embedded_hal_can::Interface for CanFifo {
+    type Id = CanId;
+    type Frame = CanFrame;
+    type Error = ();
+    type Filter = CanFilter;
+}
+
+impl embedded_hal_can::Transmitter for CanTransmitter {
+    fn transmit(
+        &mut self,
+        frame: &Self::Frame,
+    ) -> Result<Option<Self::Frame>, nb::Error<Self::Error>> {
+        let can = can_register();
+
+        for tx_idx in 0..3 {
+            let free = match tx_idx {
+                0 => can.tsr.read().tme0().bit_is_set(),
+                1 => can.tsr.read().tme1().bit_is_set(),
+                2 => can.tsr.read().tme2().bit_is_set(),
+                _ => unreachable!(),
+            };
+
+            if !free {
+                continue;
+            }
+
+            let tx = &can.tx[tx_idx];
+
+            match frame.id {
+                CanId::BaseId(id) => tx.tir.modify(|_, w| unsafe { w.stid().bits(id) }),
+                CanId::ExtendedId(id) => tx.tir.modify(|_, w| unsafe {
+                    let hb = ((id & EXID_MASK) >> 18) as u16;
+                    w.exid().bits(id).stid().bits(hb).ide().set_bit()
+                }),
+            }
+
+            if let Some(_) = frame.data() {
+                unsafe {
+                    for j in 0..frame.data.len() {
+                        let val = &frame.data[j];
+                        match j {
+                            0 => tx.tdlr.modify(|_, w| w.data0().bits(*val)),
+                            1 => tx.tdlr.modify(|_, w| w.data1().bits(*val)),
+                            2 => tx.tdlr.modify(|_, w| w.data2().bits(*val)),
+                            3 => tx.tdlr.modify(|_, w| w.data3().bits(*val)),
+                            4 => tx.tdhr.modify(|_, w| w.data4().bits(*val)),
+                            5 => tx.tdhr.modify(|_, w| w.data5().bits(*val)),
+                            6 => tx.tdhr.modify(|_, w| w.data6().bits(*val)),
+                            7 => tx.tdhr.modify(|_, w| w.data7().bits(*val)),
+                            _ => unreachable!(),
+                        }
+                    }
+
+                    tx.tdtr.modify(|_, w| w.dlc().bits(frame.data.len() as u8));
+                }
+
+                tx.tir.modify(|_, w| w.rtr().clear_bit());
+            } else {
+                tx.tir.modify(|_, w| w.rtr().set_bit());
+            }
+
+            // Request the frame be sent!
+            tx.tir.modify(|_, w| w.txrq().set_bit());
+
+            return Ok(None);
+        }
+
+        Err(nb::Error::WouldBlock)
+    }
+}
+
+impl Receiver for CanFifo {
+    fn receive(&mut self) -> Result<Self::Frame, Error<Self::Error>> {
+        let can = can_register();
```

Comment:
I would make this explicit and not hide it behind a function, as it is done everywhere else in this hal IIRC

```suggestion
        let can = unsafe { &*stm32::CAN::ptr() };
```